### PR TITLE
#30 Add missing RIPE API methods

### PR DESCRIPTION
**Pull request by 3rdvision**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/ripe-api/issues/30 |
| Decisions | - Add missing endpoint methods according to RIPE SDK method collection (check below list). <br> - Imports in alphabetical order.  |

## Added endpoint methods list
|   RIPE CORE Endpoint                                                              |   RIPE API (Python) Method                                                      |   RIPE SDK (Javascript) Method                                         |
|-----------------------------------------------------------------------------------|---------------------------------------------------------------------------------|------------------------------------------------------------------------|
|   /api/contentful/clear                                                           |   -                                                                             |   -                                                                    |
|   /api/contentful/values/<str:key>                                                |   -                                                                             |   -                                                                    |
|   /api/pushi/auth                                                                 |   -                                                                             |   -                                                                    |
|   /api/pushi/notifications                                                        |   -                                                                             |   -                                                                    |
|   /api/accounts/me                                                                |   +account_me                                                                   |   accountMe                                                            |
|   /api/accounts/me/tenancy                                                        |   +tenancy_account_me                                                           |   tenancyAccountMe                                                     |
|   /api/ping                                                                       |   +ping                                                                         |   ping                                                                 |
|   /api/geo_resolve                                                                |   +geo_resolve                                                                  |   geoResolve                                                           |
|   /api/signin                                                                     |   login                                                                         |   signin                                                               |
|   /api/signin_key                                                                 |   +login_key                                                                    |   -                                                                    |
|   /api/signin_pid                                                                 |   login_pid                                                                     |   signinPid                                                            |
|   /api/signin_admin                                                               |   login                                                                         |   signinAdmin                                                          |
|   /api/structure                                                                  |   +structure                                                                    |   -                                                                    |
|   /api/thumb/<str:model>                                                          |   +thumb                                                                        |   -                                                                    |
|   /api/translate                                                                  |   +translate                                                                    |   -                                                                    |
|   /api/compose                                                                    |   +compose                                                                      |   _getImageOptions                                                     |
|   /api/mask                                                                       |   +mask                                                                         |   _getMaskURL                                                          |
|   /api/swatch                                                                     |   +swatch                                                                       |   _getSwatchURL                                                        |
|   /api/compose/clear                                                              |   +clear_cache_compose                                                          |   -                                                                    |
|   /api/brands/<str:brand>/logo.<str:format>                                       |   logo_brand                                                                    |   getLogoUrl                                                           |
|   /api/brands/<str:brand>/models/<str:model>/mesh                                 |   +mesh_brand                                                                   |   getMesh                                                              |
|   /api/brands/<str:brand>/models/<str:model>/config                               |   config_brand                                                                  |   getConfig                                                            |
|   /api/brands/<str:brand>/models/<str:model>/spec                                 |   -                                                                             |   -                                                                    |
|   /api/brands/<str:brand>/models/<str:model>/combinations                         |   combinations_brand                                                            |   getCombinations                                                      |
|   /api/brands/<str:brand>/models/<str:model>/factory                              |   factory_brand                                                                 |   getFactory                                                           |
|   /api/brands/<str:brand>/models/<str:model>/validate                             |   validate_brand                                                                |   validateModel                                                        |
|   /api/brands/<str:brand>/models/<str:model>/logic                                |   +logic_brand                                                                  |   getLogic                                                             |
|   /api/brands/<str:brand>/models/<str:model>/logic/<str:method>                   |   logic_method_brand                                                            |   runLogic                                                             |
|   /api/brands/<str:brand>/models/<str:model>/logic/<str:method>/exists            |   logic_exists_brand                                                            |   -                                                                    |
|   /api/brands/<str:brand>/models/<str:model>/on_config                            |   (logic_method_brand variant)                                                  |   onConfig                                                             |
|   /api/brands/<str:brand>/models/<str:model>/on_part                              |   (logic_method_brand variant)                                                  |   onPart                                                               |
|   /api/brands/<str:brand>/models/<str:model>/on_initials                          |   (logic_method_brand variant)                                                  |   onInitials                                                           |
|   /api/builds                                                                     |   +get_builds                                                                   |   getBuilds                                                            |
|   /api/builds/local                                                               |   +get_local_builds                                                             |   getLocalBuilds                                                       |
|   /api/builds/<str:name>                                                          |   +get_build                                                                    |   getBuild                                                             |
|   /api/builds/<str:name>/install                                                  |   +install_build                                                                |   installBuild                                                         |
|   /api/builds/<str:name>/uninstall                                                |   +uninstall_build                                                              |   uninstallBuild                                                       |
|   /api/builds/<str:name>/update                                                   |   +update_build                                                                 |   updateBuild                                                          |
|   /api/builds/<str:name>/switch                                                   |   +switch_build                                                                 |   switchBuild                                                          |
|   /api/builds/<str:name>/artifacts                                                |   +get_build_artifacts                                                          |   getBuildArtifacts                                                    |
|   /api/builds/<str:name>/artifacts/<str:version>                                  |   +get_build_artifact                                                           |   getBuildArtifact                                                     |
|   /api/builds/<str:name>/artifacts/<str:version>/install                          |   +install_artifact                                                             |   installArtifact                                                      |
|   /api/builds/<str:name>/artifacts/<str:version>/uninstall                        |   +uninstall_artifact                                                           |   uninstallArtifact                                                    |
|   /api/builds/<str:name>/artifacts/<str:version>/switch                           |   +switch_artifact                                                              |   switchArtifact                                                       |
|   /api/builds/<str:name>/logo.<str:format>                                        |   -                                                                             |   -                                                                    |
|   /api/builds/<str:name>/fonts                                                    |   -                                                                             |   -                                                                    |
|   /api/builds/<str:name>/fonts/<regex('[a-zA-Z0-9]+'):font>.<str:format>          |   -                                                                             |   -                                                                    |
|   /api/builds/<str:name>/fonts/<regex('[a-zA-Z0-9]+'):font>                       |   -                                                                             |   -                                                                    |
|   /api/builds/<str:name>/locale                                                   |   +get_locale_model                                                             |   getLocaleModel                                                       |
|   /api/builds/<str:name>/locale.csv                                               |   -                                                                             |   -                                                                    |
|   /api/builds/<str:name>/locale/keys                                              |   +get_locale_model_keys                                                        |   getLocaleModelKeys                                                   |
|   /api/builds/<str:name>/locale/template                                          |   -                                                                             |   -                                                                    |
|   /api/builds/<str:name>/locale/<str:locale>                                      |   -                                                                             |   -                                                                    |
|   /api/builds/<str:name>/locale/<str:locale>/values                               |   -                                                                             |   -                                                                    |
|   /api/config                                                                     |   get_config                                                                    |   configGlobal                                                         |
|   /api/config/info                                                                |   info_config                                                                   |   configInfo                                                           |
|   /api/config/sku                                                                 |   sku_config                                                                    |   configSku                                                            |
|   /api/config/price                                                               |   price_config                                                                  |   getPrice                                                             |
|   /api/config/prices                                                              |   +prices_config                                                                |   getPrices                                                            |
|   /api/config/availability                                                        |   availability_config                                                           |   -                                                                    |
|   /api/config/resolve/<int:product_id>                                            |   resolve_config                                                                |   configResolve                                                        |
|   /api/config/defaults/<str:model>                                                |   defaults_model_config                                                         |   -                                                                    |
|   /api/country_groups                                                             |   +get_country_groups, +create_country_group                                    |   getCountryGroups, createCountryGroup                                 |
|   /api/country_groups/<int:id>                                                    |   +get_country_group                                                            |   getCountryGroup                                                      |
|   /api/designs                                                                    |   +get_designs                                                                  |   -                                                                    |
|   /api/designs/<str:id>                                                           |   +get_design                                                                   |   -                                                                    |
|   /api/email/budy                                                                 |   -                                                                             |   -                                                                    |
|   /api/email/notify                                                               |   -                                                                             |   -                                                                    |
|   /api/history                                                                    |   -                                                                             |   -                                                                    |
|   /api/history/clear                                                              |   -                                                                             |   -                                                                    |
|   /api/justifications                                                             |   +get_justifications                                                           |   getJustifications                                                    |
|   /api/justifications/<str:context>                                               |   +get_justification                                                            |   getJustificationsByContext                                           |
|   /api/justifications/<str:context>/<str:id>                                      |   -                                                                             |   -                                                                    |
|   /api/locale                                                                     |   +get_locale                                                                   |   locale                                                               |
|   /api/locale/bundle                                                              |   +get_locale_bundle                                                            |   localeBundle                                                         |
|   /api/locale/<str:value>                                                         |   -                                                                             |   -                                                                    |
|   /api/models/<str:model>/config                                                  |   config_model                                                                  |   -                                                                    |
|   /api/models/<str:model>/spec                                                    |   -                                                                             |   -                                                                    |
|   /api/models/<str:model>/combinations                                            |   combinations_model                                                            |   -                                                                    |
|   /api/models/<str:model>/factory                                                 |   factory_model                                                                 |   -                                                                    |
|   /api/models/<str:model>/validate                                                |   +validate_model                                                               |   -                                                                    |
|   /api/notify_infos/device_ids                                                    |   +create_device_id                                                             |   createDeviceId                                                       |
|   /api/notify_infos/device_ids/<str:id>                                           |   +remove_device_id                                                             |   removeDeviceId                                                       |
|   /api/orders                                                                     |   list_orders, create_order                                                     |   getOrders                                                            |
|   /api/orders/search                                                              |   +search_order                                                                 |   searchOrders                                                         |
|   /api/orders/<int:number>                                                        |   get_order                                                                     |   getOrder                                                             |
|   /api/orders/<int:number>/attachments                                            |   +get_attachments_order, +create_attachments_order                             |   attachmentsOrder, createAttachmentOrder                              |
|   /api/orders/<int:number>/attachments/<int:attachment_id>                        |   +attachment_order                                                             |   attachmentOrder                                                      |
|   /api/orders/<int:number>/attachments/count                                      |   -                                                                             |   -                                                                    |
|   /api/orders/<int:number>/notes                                                  |   +create_note_order                                                            |   createNoteOrder                                                      |
|   /api/orders/<int:number>/waybill                                                |   +create_waybill_order                                                         |   createWaybillOrder                                                   |
|   /api/orders/<int:number>/refresh_shipping                                       |   +refresh_shipping_order                                                       |   refreshShippingOrder                                                 |
|   /api/orders/<int:number>/log                                                    |   +log_order                                                                    |   logOrder                                                             |
|   /api/orders/<int:number>/states                                                 |   +states_order                                                                 |   statesOrder                                                          |
|   /api/orders/<int:number>/states/<int:state_id>                                  |   +state_order                                                                  |   stateOrderP                                                          |
|   /api/orders/<int:number>/states/<int:state_id>/chat                             |   +state_chat_order                                                             |   stateChatOrder                                                       |
|   /api/orders/<int:number>/states/<int:state_id>/chat/lines                       |   +state_chat_lines_order                                                       |   stateChatLinesOrder                                                  |
|   /api/orders/<int:number>/states/<int:state_id>/chat/lines/count                 |   +state_chat_lines_count_order                                                 |   stateChatLinesCountOrder                                             |
|   /api/orders/<int:number>/states/<int:state_id>/attachments                      |   +state_attachments_order                                                      |   stateAttachmentsOrder                                                |
|   /api/orders/<int:number>/states/<int:state_id>/attachments/count                |   +state_attachments_count_order                                                |   stateAttachmentsCountOrder                                           |
|   /api/orders/<int:number>/states/<int:state_id>/attachments                      |   +state_attachments_order, +state_create_attachments_order                     |   stateAttachmentsOrder, stateCreateAttachmentOrder                    |
|   /api/orders/<int:number>/states/<int:state_id>/attachments/<int:attachment_id>  |   +state_attachment_order                                                       |   stateAttachmentOrder                                                 |
|   /api/orders/<int:number>/factory                                                |   -                                                                             |   -                                                                    |
|   /api/orders/<int:number>/price                                                  |   -                                                                             |   -                                                                    |
|   /api/orders/<int:number>/transport                                              |   +get_transport_order                                                          |   getTransportOrder                                                    |
|   /api/orders/<int:number>/report                                                 |   -                                                                             |   _getOrderReportURL                                                   |
|   /api/orders/<int:number>/report.pdf                                             |   report_pdf                                                                    |   _getOrderReportPDFURL                                                |
|   /api/orders/<int:number>/report.png                                             |   -                                                                             |   _getOrderReportPNGURL                                                |
|   /api/orders/<int:number>/image                                                  |   -                                                                             |   -                                                                    |
|   /api/orders/<int:number>/image.token                                            |   -                                                                             |   -                                                                    |
|   /api/orders/<int:number>/subscription                                           |   +get_subscription_order, +subscribe_order, +unsubscribe_order                 |   getOrderSubscription, subscribeOrder, unsubscribeOrderP              |
|   /api/orders/<int:number>/report_url                                             |   -                                                                             |   -                                                                    |
|   /api/orders/shipping                                                            |   -                                                                             |   -                                                                    |
|   /api/orders/import                                                              |   +import_order                                                                 |   importOrder                                                          |
|   /api/orders/import.pdf                                                          |   -                                                                             |                                                                        |
|   /api/orders/pre_customization                                                   |   +precustomization_order                                                       |   precustomizationOrder                                                |
|   /api/orders/farfetch                                                            |   +create_farfetch_order                                                        |   -                                                                    |
|   /api/orders/budy                                                                |   -                                                                             |   -                                                                    |
|   /api/orders/budy/notify                                                         |   -                                                                             |   -                                                                    |
|   /api/plugins                                                                    |   -                                                                             |   -                                                                    |
|   /api/plugins                                                                    |   -                                                                             |   -                                                                    |
|   /api/plugins/<str:name>/artifacts/<str:version>                                 |   -                                                                             |   -                                                                    |
|   /api/products                                                                   |   -                                                                             |   -                                                                    |
|   /api/products/<str:brand>/<str:model>/config                                    |   -                                                                             |   -                                                                    |
|   /api/profiles                                                                   |   +get_profiles                                                                 |   getProfiles                                                          |
|   /api/profiles/<str:name>                                                        |   +get_profile                                                                  |   getProfile                                                           |
|   /api/sizes                                                                      |   get_sizes                                                                     |   getSizes                                                             |
|   /api/sizes/size_to_native                                                       |   size_to_native                                                                |   sizeToNative                                                         |
|   /api/sizes/size_to_native_b                                                     |   size_to_native_b                                                              |   sizeToNativeB                                                        |
|   /api/sizes/native_to_size                                                       |   native_to_size                                                                |   nativeToSize                                                         |
|   /api/sizes/native_to_size_b                                                     |   native_to_size_b                                                              |   nativeToSizeB                                                        |
|   /api/sizes/size_to_locale                                                       |   +size_to_locale                                                               |   sizeToLocale                                                         |
|   /api/sizes/size_to_locale_b                                                     |   +size_to_locale_b                                                             |   sizeToLocaleB                                                        |
|   /api/sizes/native_to_locale                                                     |   +native_to_locale                                                             |   nativeToLocale                                                       |
|   /api/sizes/native_to_locale_b                                                   |   +native_to_locale_b                                                           |   nativeToLocaleB                                                      |
|   /api/sizes/locale_to_native                                                     |   locale_to_native                                                              |   -                                                                    |
|   /api/sizes/locale_to_native_b                                                   |   locale_to_native_b                                                            |   -                                                                    |
|   /api/skus                                                                       |   +get_skus, +create_sku                                                        |   getSkus, createSku                                                   |
|   /api/skus/<int:id>                                                              |   +get_sku                                                                      |   getSku                                                               |
|   /api/skus/<int:id>/validate                                                     |   +validateSku                                                                  |   validate_sku                                                         |
|   /api/webhooks                                                                   |   -                                                                             |   -                                                                    |
|   /api/webhooks/<int:id>                                                          |   -                                                                             |   -                                                                    |
|   /api/availability_rules                                                         |   +get_availability_rules, +create_availability_rule                            |   getAvailabilityRulesP, createAvailabilityRule                        |
|   /api/availability_rules/<int:id>                                                |   +get_availability_rule, +update_availability_rule, +delete_availability_rule  |   getAvailabilityRule, updateAvailabilityRule, deleteAvailabilityRule  |
|   /api/factory_rules                                                              |   +get_factory_rules, +create_factory_rule                                      |   getFactoryRules, createFactoryRule                                   |
|   /api/factory_rules/<int:id>                                                     |   +get_factory_rule, +update_factory_rule, +delete_factory_rule                 |   getFactoryRuleP, updateFactoryRule, deleteFactoryRule                |
|   /api/letter_rules                                                               |   +get_letter_rules, +create_letter_rule                                        |   getLetterRules, createLetterRule                                     |
|   /api/letter_rules/<int:id>                                                      |   +get_letter_rule, +update_letter_rule, +delete_letter_rule                    |   getLetterRule, updateLetterRule, deleteLetterRule                    |
|   /api/price_rules                                                                |   +get_price_rules, +create_price_rule                                          |   getPriceRules, createPriceRule                                       |
|   /api/price_rules/<int:id>                                                       |   +get_price_rule, +update_price_rule, +delete_price_rule                       |   getPriceRule, updatePriceRule, deletePriceRule                       |
|   /api/transport_rules                                                            |   +get_transport_rules, +create_transport_rule                                  |   getTransportRules, createTransportRule                               |
|   /api/transport_rules/<int:id>                                                   |   +get_transport_rule, +update_transport_rule, +delete_transport_rule           |   getTransportRule, updateTransportRule, deleteTransportRule           |

### CODE REVIEW

**Comment by NFSS10** [comment]
@3rdvision Why is this in draft?

File: src/ripe/transport_rule.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+class TransportRuleAPI(object):
+
+    def get_transport_rules(self, *args, **kwargs):
```
Comment (joamag) [inline src/ripe/transport_rule.py:6]: list

File: src/ripe/sku.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+class SkuAPI(object):
+
+    def get_skus(self):
```
Comment (joamag) [inline src/ripe/sku.py:6]: list_skus

File: setup.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 
 setuptools.setup(
     name = "ripe-api",
-    version = "0.4.0",
+    version = "0.5.0",
```
Comment (joamag) [inline setup.py:9]: we should not change versions in feature PRs

File: src/ripe/availability_rule.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+class AvailabilityRuleAPI(object):
+    
+    def get_availability_rules(self, *args, **kwargs):
```
Comment (joamag) [inline src/ripe/availability_rule.py:6]: list_availability_rules

**Review by joamag — CHANGES_REQUESTED** [review]

File: src/ripe/availability_rule.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+# -*- coding: utf-8 -*-
+
+class AvailabilityRuleAPI(object):
+    
+    def list_availability_rules(self, *args, **kwargs):
+        url = self.base_url + "availability_rules"
+        contents = self.get(url, **kwargs)
+        return contents
+
+    def get_availability_rule(self, id):
```
Comment (joamag) [inline src/ripe/availability_rule.py:11]: the order of the methods should be LCRUD

**Review by joamag — CHANGES_REQUESTED** [review]

File: src/ripe/order.py [hunk 6]
```diff
@@ -27,6 +27,246 @@ def get_order(self, number):
...
+        url = self.base_url + "orders/%d/states/%d/attachments" % (number, state_id)
+        contents = self.post(url, data_j = file)
+        return contents
+
+    def state_attachment_order(self, number, state_id, attachment_id):
+        url = self.base_url + "orders/%d/states/%d/attachments/%d" % (number, state_id, attachment_id)
+        contents = self.get(url)
+        return contents
+
+    def quality_assure_order(
```
Comment (joamag) [inline src/ripe/order.py:120]: Please be aware of proper sorting of methods, the quality assure should be together with the ready, produce, etc. methods

**Review by joamag — CHANGES_REQUESTED** [review]

File: src/ripe/build.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/python
```
Comment (joamag) [inline src/ripe/build.py:1]: remove this API for now, we don't want the build API to get exposed for now

**Review by joamag — CHANGES_REQUESTED** [review]

File: src/ripe/base.py [hunk 8]
```diff
@@ -38,6 +64,7 @@ def __init__(self, *args, **kwargs):
         self.admin = kwargs.get("admin", self.admin)
         self.session_id = kwargs.get("session_id", None)
         self.token = kwargs.get("token", None)
+        self.key = kwargs.get("key", None)
```
Comment (joamag) [inline src/ripe/base.py:67]: we don't want to suport this approach, please remove key references

**Review by joamag — CHANGES_REQUESTED** [review]

**Review by joamag — APPROVED** [review]

**Pull request merged**